### PR TITLE
[ML] Implement v1 ML pipeline – Trailer character predictions (age, gender, ethnicity)

### PR DESCRIPTION
**Pull request by Julien-Commes**
This PR introduces the first version of the ML pipeline for analyzing characters in movie trailers. The pipeline extracts faces and people, applies filtering, classifies attributes (age, gender, ethnicity), and clusters faces to group identical characters.

**Key additions:**

`vision_detection.py` – Extracts faces and people from frames
`filter_detections.py` – Filters "good candidates" based on predefined criteria
`vision_classifier.py` – Classifies faces (age, gender, ethnicity)
`faces_clustering.py` – Clusters faces to identify the same characters
`utils.py`, `main.py`, `model_architectures.py` – Integrates the above steps for end-to-end inference from a trailer URL

**Comments:** This branch was rebased onto main to ensure a linear commit history. As a result, all commits appear with the same timestamp.

### CODE REVIEW

File: ml-image/main.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    ### Detect faces in the video
+    detections = vision_dectector.crop_areas_of_interest(frames, H_original, W_original, area_type = 'face', batch_size = 256)
+    print("Faces detected", datetime.now() - start_time)
+
+    ### Filter detections
+    #total_area = H_original * W_original
+    effective_height = W_original / 2.4
+    effective_area = effective_height * W_original
+
+    filters = DetectionFilter(simple_filters = [filter_by_area, filter_by_confidence], complex_filters = [filter_by_sharpness, filter_by_pose], area_type = 'face', total_area = effective_area, min_area = 0.01, min_sharpness = 200.0, min_conf = 0.6)
```
Comment (AudreyDuran) [inline ml-image/main.py:46]: pr les hyperparam ici aussi ce serait pas mal de définir des constantes qqpart (utile pr un grid search ensuite) - s'applique pr ttes les constantes définies dans le code

File: ml-image/main.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    else:
+        source = sys.argv[1]
+        if source[:4] == "http":
+            print(f"Looking for data at: {source}")
+            data = get_data_from_url(source)
+        else:
+            print(f"Looking for data in file: {source}")
+            data = get_data_from_file(source, filetype="trailer")
+
+    print("Data loaded", datetime.now() - start_time)
```
Comment (AudreyDuran) [inline ml-image/main.py:25]: ce serait pas mal d'avoir quelques methodes ici, par exemple: `load_data`, `detect_faces`, `classify_faces_bb`, `filter_faces` etc

File: ml-image/scripts/faces_clustering.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        if len(embeddings) == 0:
+            return np.array([])
+
+        embeddings = torch.cat(embeddings)  # Concatenate all embeddings
+        embeddings_array = embeddings.numpy() # Convert to numpy array for use in clustering algorithms
+
+        return embeddings_array
+
+class Faces_clustering():
+    def __init__(self, model = "chinese_whispers", **kwargs):
```
Comment (AudreyDuran) [inline ml-image/scripts/faces_clustering.py:56]: peut etre mettre directement threshold plutot que kwargs ?
```suggestion
    def __init__(self, model = "chinese_whispers", threshold: float,  **kwargs):
            self.model = model
            self.threshold = threshold
```

File: ml-image/scripts/faces_clustering.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        if len(embeddings) == 0:
+            return np.array([])
+
+        embeddings = torch.cat(embeddings)  # Concatenate all embeddings
+        embeddings_array = embeddings.numpy() # Convert to numpy array for use in clustering algorithms
+
+        return embeddings_array
+
+class Faces_clustering():
```
Comment (AudreyDuran) [inline ml-image/scripts/faces_clustering.py:55]: si on veut suivre pep8 il faut du CamelCase ici (`class FacesClustering`)

File: ml-image/scripts/faces_clustering.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                    case "majority":
+                        aggregated_age = statistics.median([person["age"] for person in persons])
+                        aggregated_gender = statistics.mode([person["gender"] for person in persons])
+                        aggregated_ethnicity = statistics.mode([person["ethnicity"] for person in persons])
+                        occurence = len(persons)
+                        area_occupied = sum([np.abs(x1 - x2) * np.abs(y1 - y2) for (x1, y1, x2, y2) in [person["bbox"] for person in persons]])
+                        persons_id = [person["person_id"] for person in persons]
+                        aggregated_persons.append({"age": aggregated_age, "gender": aggregated_gender, "ethnicity": aggregated_ethnicity, "occurence": occurence, "area occupied": area_occupied, "persons_id": persons_id, "label": final_label})
+                    case _:
+                        raise ValueError("Method not supported")
```
Comment (AudreyDuran) [inline ml-image/scripts/faces_clustering.py:100]:
```suggestion
                        raise ValueError(f"Method  {method} not supported")
```

File: ml-image/scripts/filter_detections.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import cv2
+import numpy as np
+from typing import List, Dict, Callable
+import mediapipe as mp
+
+class DetectionFilter:
+    def __init__(self, simple_filters: List[Callable[[Dict], bool]], complex_filters: List[Callable[[Dict], bool]], area_type : str = 'face', **kwargs) -> None:
```
Comment (AudreyDuran) [inline ml-image/scripts/filter_detections.py:7]: quelle est la différence entre les filtres simples et complexes?

File: ml-image/scripts/filter_detections.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        self.simple_filters = simple_filters
+        self.complex_filters = complex_filters
+        self.area_type = area_type
+        self.kwargs = kwargs
+
+    def apply(self, detections: List[Dict]) -> List[Dict]:
+        filtered_detections = [det for det in detections if all(f(det, **self.kwargs) for f in self.simple_filters)]
+
+        for det in filtered_detections:
+            det["sharpness"] = compute_sharpness(det[f"cropped_{self.area_type}"])  # On ne calcule que pour celles qui restent
```
Comment (AudreyDuran) [inline ml-image/scripts/filter_detections.py:17]: à ne calculer que si la sharpness fait partie des complex_filters?

File: ml-image/scripts/filter_detections.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+def filter_by_area(det: Dict, **kwargs) -> bool:
+    min_area = kwargs.get("min_area", 0.01)
+    max_area = kwargs.get("max_area", 0.5)
+    x1, y1, x2, y2 = det["bbox"]
+    total_area = kwargs.get("total_area", 1.0)
+    area = np.abs(x2 - x1) * np.abs(y2 - y1) / total_area
+    return min_area <= area <= max_area 
+
+def filter_by_confidence(det: Dict, **kwargs) -> bool:
```
Comment (AudreyDuran) [inline ml-image/scripts/filter_detections.py:29]: je renommerais genre `validates_confidence_filter` comme ça retourne un bool (idem pr les autres methodes)

File: ml-image/scripts/filter_detections.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    return cv2.Laplacian(gray, cv2.CV_64F).var()
+
+def filter_by_sharpness(det: Dict, **kwargs) -> bool:
+    min_sharpness = kwargs.get("min_sharpness", 35.0)
+    return det["sharpness"] >= min_sharpness
+
+mp_face_mesh = mp.solutions.face_mesh
+face_mesh = mp_face_mesh.FaceMesh(static_image_mode=True, refine_landmarks=True)
+
+def get_face_landmarks(image: np.ndarray) -> tuple:
```
Comment (AudreyDuran) [inline ml-image/scripts/filter_detections.py:44]: je verrais bien une classe pr chaque filtre (qui hériteraient d'une meme classe abstraite) mais peut etre pas la priorité du projet :p

File: ml-image/scripts/model_architectures.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
+import torch
```
Comment (AudreyDuran) [inline ml-image/scripts/model_architectures.py:1]:
```suggestion
"""Classifier models zoo."""

import torch
```

File: ml-image/scripts/utils.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
+import requests
```
Comment (AudreyDuran) [inline ml-image/scripts/utils.py:1]: je renommerais dataset.py plutot que utils.py mais c'est du détail

File: ml-image/scripts/utils.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+import cv2
+import numpy as np
+
+from bs4 import BeautifulSoup
+from PIL import Image
+from torch.utils.data import Dataset
+
+class ImageDataset(Dataset):
+    # Dataset builder for array images for specific PyTorch model compliance
+    def __init__(self, image_array, transform):
```
Comment (AudreyDuran) [inline ml-image/scripts/utils.py:13]: pas besoin d'une liste de transform, une suffit?

File: ml-image/scripts/utils.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        for person in frame_persons:
+            if person["frame_id"] == frame_index:
+                bbox = [int(coord) for coord in person["bbox"]]  # Ensure bbox coordinates are integers
+                for video_person in video_persons:
+                    if person["person_id"]  in video_person["persons_id"]:
+                        match video_person["gender"]:
+                            case "male":
+                                color = (0, 0, 255)
+                            case "female":
+                                color = (255, 0, 0)
```
Comment (AudreyDuran) [inline ml-image/scripts/utils.py:127]: ça a changé? il me semblait que c'était l'inverse dans les exemples que vous nous aviez montrés, rouge les hommes et bleu les femmes

File: ml-image/scripts/utils.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+def store_predictions_on_video(frames: np.ndarray, video_persons: list, frame_persons: list, fps: int = 25, output_name: str = 'predictions_trailer.avi') -> None:
+    """
+    Draw predictions on video frames and store the result in a new video file
+    """
+    modified_frames = draw_predictions_on_video(frames, video_persons, frame_persons)
+
+    # Define the codec and create VideoWriter object
+    height, width, _= modified_frames[0].shape
+    output_path = os.path.join('example', output_name)
```
Comment (AudreyDuran) [inline ml-image/scripts/utils.py:144]: fyi la librairie pathlib est vrmt pratique pour gérer les paths

**Review by AudreyDuran — COMMENTED** [review]
top merci beaucoup pour votre travail c'est cool ! j'ai mis pas mal de commentaires mais libre à vous de les ignorer, c'est surtout de l'ordre du style et mineur:)

File: ml-image/.gitignore [hunk 15]
```diff
@@ -0,0 +1,5 @@
+models/*
+example/*
+grid_search_generator.py
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/.gitignore:3]: On pourra l'ajouter plus tard mais pour l'instant c'est une v0 (j'avais écrit le code la veille).

File: ml-image/scripts/utils.py (hunk 13 above)
Comment (Julien-Commes) [reply ml-image/scripts/utils.py:127]: Pas de changement, le color space est en BGR.

File: ml-image/scripts/filter_detections.py (hunk 7 above)
Comment (Julien-Commes) [reply ml-image/scripts/filter_detections.py:17]: A priori je mettrais toujours sharpness en complex_filters, mais avec du recul je pense que ce sera mieux de directement mettre l'appel à la méthode `compute_sharpness` dans `filter_by_sharpness`

File: ml-image/scripts/utils.py (hunk 13 above)
Comment (AudreyDuran) [reply ml-image/scripts/utils.py:127]: ah ok! merci

File: ml-image/main.py (hunk 2 above)
Comment (Julien-Commes) [reply ml-image/main.py:25]: Pour confirmer que j'ai bien compris, tu veux dire regrouper les quelques lignes de chaque étape dans des méthodes plutôt que de les lister avec un commentaire avant type `### Extract frames from video` ?

File: ml-image/scripts/utils.py (hunk 12 above)
Comment (Julien-Commes) [reply ml-image/scripts/utils.py:13]: Pas sûr de comprendre. De la façon dont on l'utilise on donne un transform.Compose du style : 
`inference_transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Resize(size=(64,64)),
            transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]),
            ])`

File: ml-image/main.py (hunk 2 above)
Comment (AudreyDuran) [reply ml-image/main.py:25]: yes exactement mais c'est mineur dans tous les cas:)

File: ml-image/scripts/filter_detections.py (hunk 6 above)
Comment (AudreyDuran) [reply ml-image/scripts/filter_detections.py:7]: ça roule merci!

File: ml-image/scripts/filter_detections.py (hunk 7 above)
Comment (AudreyDuran) [reply ml-image/scripts/filter_detections.py:17]: je suis d'accord

File: ml-image/main.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
+from scripts import VisionDetection, VisionClassifier, Embedding_model, Faces_clustering, DetectionFilter, filter_by_area, filter_by_confidence, filter_by_sharpness, filter_by_pose, get_data_from_file, get_data_from_url, store_predictions_on_video, frame_capture
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+from datetime import datetime
+import torch
+
+import sys
+
+def main() :
+    start_time = datetime.now()
+
+    ### Get data from url
+    if len(sys.argv) < 2:
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    vision_dectector = VisionDetection(device = device, num_cpu_threads = 8)
+    classifier = VisionClassifier(device = device)
+    embedding_model = Embedding_model(model = "facenet", device = device)
+    
+    ### Detect faces in the video
+    detections = vision_dectector.crop_areas_of_interest(frames, H_original, W_original, area_type = 'face', batch_size = 256)
+    print("Faces detected", datetime.now() - start_time)
+
+    ### Filter detections
+    #total_area = H_original * W_original
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+    print("Data loaded", datetime.now() - start_time)
+    ### Extract frames from video
+    frames = frame_capture(data["trailer_path"])
+    print("Frames extracted", datetime.now() - start_time)
+    H_original, W_original = frames[0].shape[:2]
+
+    ### Initialize models
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    vision_dectector = VisionDetection(device = device, num_cpu_threads = 8)
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    classifier = VisionClassifier(device = device)
+    embedding_model = Embedding_model(model = "facenet", device = device)
+    
+    ### Detect faces in the video
+    detections = vision_dectector.crop_areas_of_interest(frames, H_original, W_original, area_type = 'face', batch_size = 256)
```
[the remark this answers is not in the document]
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py (hunk 1 above)
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/main.py (hunk 2 above)
Comment (Julien-Commes) [reply ml-image/main.py]: - [X] Corrigé dans le dernier commit

File: ml-image/scripts/faces_clustering.py (hunk 4 above)
Comment (Julien-Commes) [reply ml-image/scripts/faces_clustering.py]: - [X] Corrigé dans le dernier commit. Merci pour le tips

File: ml-image/scripts/filter_detections.py (hunk 7 above)
Comment (Julien-Commes) [reply ml-image/scripts/filter_detections.py]: - [X] Corrigé dans le dernier commit

File: ml-image/scripts/filter_detections.py (hunk 8 above)
Comment (Julien-Commes) [reply ml-image/scripts/filter_detections.py]: - [X] Corrigé dans le dernier commit

File: ml-image/scripts/filter_detections.py (hunk 9 above)
Comment (Julien-Commes) [reply ml-image/scripts/filter_detections.py:42]: Ajouté à la ToDo pour une version affinée

File: ml-image/scripts/utils.py (hunk 14 above)
Comment (Julien-Commes) [reply ml-image/scripts/utils.py:144]: Ajouté à la ToDo pour une version affinée

File: ml-image/scripts/utils.py [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+    def __len__(self):
+        return len(self.image)
+
+    def __getitem__(self, idx):
+        image = self.image[idx]
+        if self.transform:
+            return self.transform(image)
+        raise ValueError("No transformation provided")
+
+def get_data_from_url(url: str,
```
Comment (fitzfoufou) [inline ml-image/scripts/utils.py:26]: Hello @Julien-Commes, ce serait cool d'isoler la récupération des données allocine dans une classe a part.
Si tu pouvais rebase ta branche sur main, tu pourrais voir dans `database/mubi/` un exemple d'implementation. Si tu pouvais mettre cette classe dans un dossier `database/allocine/` ce serait super, parce que nous on va aussi en avoir besoin cote data.
merci d'avance

**Review by fitzfoufou — CHANGES_REQUESTED** [review]
